extensions/MikeDEV/peerjs.js

```diff
@@ -0,0 +1,962 @@
+
+      const lock_id = "mikedevpeerjs_" + ID + "_" + CHANNEL;
+      await navigator.locks.request(lock_id, { ifAvailable: true }, () => {
+        // Create a new channel with PeerJS
+        const conn = this.dataConnections.get(ID);
+        const id = conn.idCounter++;
+
+        // Since PeerJS doesn't natively support multiple channels, we have to create a new one manually
+        const chan = conn.peerConnection.createDataChannel(CHANNEL, {
+          ordered: ORDERED,
+          negotiated: true,
+          id: id,
+        });
+
+        chan.onopen = () => {
+          this.handleChannelOpen(conn, chan);
+        };
+
+        chan.onclose = () => {
+          this.handleChannelClose(conn, chan);
+        };
+
+        chan.onerror = (err) => {
+          this.handleChannelError(conn, chan, err);
+        };
+
+        chan.onmessage = async (msg) => {
+          await this.handleChannelData(conn, chan, JSON.parse(msg.data));
+        };
+
+        // Store channel reference
+        conn.channels.set(CHANNEL, {
+          data: "",
+          chan: chan,
+        });
+
+        // Tell the peer about the new channel
+        conn.channels.get("default").chan.send(
+          JSON.stringify({
+            opcode: "NEW_CHAN",
+            payload: {
+              id: id,
+              label: CHANNEL,
+              ordered: ORDERED,
+            },
+          })
+        );
+      });
+    }
+
+    async sendMessageToPeer({ MESSAGE, ID, CHANNEL }) {
+      ID = Scratch.Cast.toString(ID);
+      if (!this.peer) return;
+      if (!this.dataConnections.has(ID)) return;
+      const conn = this.dataConnections.get(ID);
+      if (!conn.channels.has(CHANNEL)) return;
+      return new Promise((resolve) => {
+        const packet = JSON.stringify({
+          opcode: "P_MSG",
+          payload: MESSAGE,
+        });
+        conn.channels.get(CHANNEL).chan.send(packet);
+        resolve();
+      });
+    }
+
+    disconnectPeer() {
+      if (!this.peer) return;
+      if (!this.peer.disconnected) this.peer.disconnect();
+    }
+
+    reconnectPeer() {
+      if (!this.peer) return;
+      if (this.peer.disconnected) this.peer.reconnect();
+    }
+
+    destroyPeer() {
+      if (!this.peer) return;
+      if (!this.peer.destroyed) this.peer.destroy();
+    }
+
+    isPeerConnected() {
+      if (!this.peer) return false;
+      return !this.peer.disconnected && !this.peer.destroyed;
+    }
+
+    whenPeerGetsMessage({ ID, CHANNEL }) {
+      return this.doesPeerHaveChannel({
+        ID: Scratch.Cast.toString(ID),
+        CHANNEL,
+      });
+    }
+
+    whenSpecificPeerConnects({ ID }) {
+      return this.isOtherPeerConnected({ ID: Scratch.Cast.toString(ID) });
+    }
+
+    whenSpecificPeerDisconnects({ ID }) {
+      return !this.isOtherPeerConnected({ ID: Scratch.Cast.toString(ID) });
+    }
+
+    doesPeerHaveChannel({ ID, CHANNEL }) {
+      if (!this.peer) return false;
+      const conn = this.dataConnections.get(ID);
```

Comment:
Small oversight during fixes, you forgot to add the `Cast` stuff here.

Reply:
Realized it while working on merging codebases with CL5. Fix incoming...